abi-bindings/go/packer/packer_test.go

```diff
@@ -0,0 +1,204 @@
+						continue
+					}
+					// Lookup the object by name and cast it to types.Named
+					typeName := typeSpec.Name.Name
+					typeObj := pkg.Types.Scope().Lookup(typeName)
+					if typeObj == nil {
+						continue
+					}
+					named, ok := typeObj.Type().(*types.Named)
+					if !ok {
+						continue
+					}
+					// Finally check if the named struct we have implements the interface
+					if types.Implements(types.NewPointer(named), iface) {
+						allImplementers = append(allImplementers, named.Obj().Name())
+					}
+				}
+			}
+		}
+	}
+	return allImplementers
+}
+
+func TestExhaustivePacking(t *testing.T) {
+	implementers := findAllImplementers(t)
+	for _, structName := range implementers {
+		packerType, ok := packerTypes[structName]
+		require.True(
+			t,
+			ok,
+			fmt.Sprintf("Struct %s not found in packer_test.go packerTypes map. Add it to check for exhaustiveness", structName),
+		)
+
+		randomizedStruct, err := randomizeStruct(packerType)
+		require.NoError(t, err)
+
+		v1 := randomizedStruct.(ABIPacker)
+		encoded, err := v1.Pack()
+		require.NoError(t, err)
+		v2 := reflect.New(reflect.TypeOf(v1).Elem()).Interface().(ABIPacker)
+		err = v2.Unpack(encoded)
+		require.NoError(t, err)
+		require.Equal(t, v1, v2)
+	}
+}
+
+func randomizeStruct(packerStruct interface{}) (interface{}, error) {
+	v := reflect.ValueOf(packerStruct).Elem()
+
+	for i := 0; i < v.NumField(); i++ {
+		field := v.Field(i)
+		err := randomizeField(field)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return packerStruct, nil
+}
+
+func randomizeField(field reflect.Value) error {
+	fieldType := field.Type()
+	switch fieldType.Kind() {
+	// The list only includes types supported by abigen
+	case reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		field.SetInt(rand.Int63())
+	case reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		field.SetUint(rand.Uint64())
+	case reflect.Bool:
+		field.SetBool(rand.Intn(2) == 1)
+	case reflect.String:
+		field.SetString(randomString(rand.Intn(256)))
+	case reflect.Slice:
+		n := rand.Intn(256)
+		field.Set(reflect.MakeSlice(fieldType, n, n))
+		for i := 0; i < field.Len(); i++ {
+			err := randomizeField(field.Index(i))
+			if err != nil {
+				return fmt.Errorf("Error randomizing slice: %v", err)
+			}
+		}
+	case reflect.Struct:
+		randomizeStruct(field.Addr().Interface())
+	case reflect.Array:
+		elemType := fieldType.Elem()
+		// arrays other than byte arrays get converted to slices by abigen
+		if elemType.Kind() != reflect.Uint8 {
+			return fmt.Errorf("Unsupported array type %s. Only byte arrays are supported.", elemType.Kind().String())
+		}
+		for j := 0; j < field.Len(); j++ {
+			field.Index(j).SetUint(uint64(rand.Intn(256)))
+		}
+
+	case reflect.Pointer:
+		if fieldType.Elem() == reflect.TypeOf(big.Int{}) {
+			newBigInt := big.NewInt(rand.Int63())
+			field.Set(reflect.ValueOf(newBigInt))
+		} else {
+			return fmt.Errorf("Unsupported pointer type %s", fieldType.Elem().String())
```

Comment:
Where is it documented that big.Int is the only supported pointer type?

Reply:
Honestly, initially ChatGPT gave me a type conversion table but I confirmed it by looking at [Solidity abi-spec ](https://docs.soliditylang.org/en/latest/abi-spec.html) and the abi-gen code itself. All solidity types should be accounted for but if this changes in the future we will get an error in tests and it should be an easy fix to account for additional supported types.